fluent-syntax/src/parser.js

```diff
@@ -415,80 +415,83 @@ export default class FluentParser {
   }
 
   getPattern(ps) {
-    let buffer = '';
     const elements = [];
-    let firstLine = true;
+    ps.skipInlineWS();
 
-    if (this.withSpans) {
-      var spanStart = ps.getIndex();
+    // Special-case: trim leading whitespace and newlines.
+    if (ps.isPeekNextLinePattern()) {
+      ps.skipBlankLines();
+      ps.skipInlineWS();
     }
 
     let ch;
     while ((ch = ps.current())) {
-      if (ch === '\n') {
-        if (firstLine && buffer.length !== 0) {
-          break;
-        }
 
+      // The end condition for getPattern's while loop is a newline
+      // which is not followed by a valid pattern continuation.
+      if (ch === '\n' && !ps.isPeekNextLinePattern()) {
+        break;
+      }
+
+      if (ch === '{') {
+        const element = this.getPlaceable(ps);
+        elements.push(element);
+      } else {
+        const element = this.getTextElement(ps);
+        elements.push(element);
+      }
+    }
+
+    return new AST.Pattern(elements);
+  }
+
+  getTextElement(ps) {
+    let buffer = '';
+
+    let ch;
+    while ((ch = ps.current())) {
+
+      if (ch === '{') {
+        return new AST.TextElement(buffer);
+      }
+
+      if (ch === '\n') {
         if (!ps.isPeekNextLinePattern()) {
-          break;
+          return new AST.TextElement(buffer);
         }
 
         ps.next();
-        ps.skipLineWS();
+        ps.skipInlineWS();
 
-        if (!firstLine) {
-          buffer += ch;
-        }
-        firstLine = false;
+        // Add the new line to the buffer
+        buffer += ch;
         continue;
-      } else if (ch === '\\') {
-        const ch2 = ps.peek();
+      }
+
+      if (ch === '\\') {
+        const ch2 = ps.next();
+
         if (ch2 === '{' || ch2 === '"') {
           buffer += ch2;
         } else {
           buffer += ch + ch2;
         }
-        ps.next();
-      } else if (ch === '{') {
-        ps.next();
-
-        ps.skipLineWS();
-
-        if (buffer.length !== 0) {
-          const text = new AST.TextElement(buffer);
-          if (this.withSpans) {
-            text.addSpan(spanStart, ps.getIndex());
-          }
-          elements.push(text);
-        }
-
-        buffer = '';
-
-        elements.push(this.getExpression(ps));
-
-        ps.expectChar('}');
-
-        if (this.withSpans) {
-          spanStart = ps.getIndex();
-        }
 
-        continue;
       } else {
         buffer += ps.ch;
       }
+
       ps.next();
     }
 
-    if (buffer.length !== 0) {
-      const text = new AST.TextElement(buffer);
-      if (this.withSpans) {
-        text.addSpan(spanStart, ps.getIndex());
-      }
-      elements.push(text);
-    }
+    return new AST.TextElement(buffer);
+  }
 
-    return new AST.Pattern(elements);
+  getPlaceable(ps) {
+    ps.expectChar('{');
+    const expression = this.getExpression(ps);
```

Comment:
`getExpression` should know how to handle nested placeables. I filed #66 for this.